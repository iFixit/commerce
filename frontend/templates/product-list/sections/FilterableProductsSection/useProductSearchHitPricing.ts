import { useAuthenticatedUser } from '@ifixit/auth-sdk';
import { computeDiscountPercentage, Money } from '@ifixit/helpers';
import { ProductSearchHit } from '@models/product-list';
import isEmpty from 'lodash/isEmpty';

export type ProductSearchHitPricing = {
   price: Money;
   compareAtPrice: Money | undefined;
   isDiscounted: boolean;
   percentage: number;
   proPricesByTier: Record<string, Money> | undefined;
};

export function useProductSearchHitPricing(
   product: ProductSearchHit
): ProductSearchHitPricing {
   const user = useAuthenticatedUser();

   let proTierPrice: number | null = null;
   const discountTier = user.data?.discountTier ?? null;
   if (discountTier) {
      const defaultVariantPrice =
         product.price_tiers?.[discountTier]?.default_variant_price;
      proTierPrice =
         defaultVariantPrice == null
            ? null
            : typeof defaultVariantPrice === 'string'
            ? parseFloat(defaultVariantPrice)
            : defaultVariantPrice;
   }

   const price = Math.min(proTierPrice ?? Infinity, product.price_float);
   const compareAtPrice = product.compare_at_price ?? product.price_float;

   const percentage = computeDiscountPercentage(
      price * 100,
      compareAtPrice * 100
   );
   const isDiscounted = percentage > 0;

   const proPricesByTier =
      product.price_tiers && !isEmpty(product.price_tiers)
         ? Object.fromEntries(
              Object.entries(product.price_tiers).map(([tier, price]) => [
                 tier,
                 {
                    amount: price.default_variant_price,
                    currencyCode: 'usd',
                 },
              ])
           )
         : undefined;

   return {
      price: { amount: price, currencyCode: 'usd' },
      compareAtPrice: { amount: compareAtPrice, currencyCode: 'usd' },
      isDiscounted,
      percentage,
      proPricesByTier,
   };
}
