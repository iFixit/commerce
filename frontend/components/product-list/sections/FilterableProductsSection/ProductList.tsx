import placeholderImageUrl from '@assets/images/no-image-fixie.jpeg';
import {
   Badge,
   Box,
   Button,
   Divider,
   Flex,
   Heading,
   HStack,
   LinkBox,
   LinkOverlay,
   Stack,
   Text,
   VStack,
} from '@chakra-ui/react';
import { ThemeTypings } from '@chakra-ui/styled-system';
import { Rating } from '@components/ui';
import { useAppContext } from '@ifixit/app';
import { ProductSearchHit } from '@models/product-list';
import { IfixitImage } from '@components/ifixit-image';
import * as React from 'react';
import { useProductSearchHitPricing } from './useProductSearchHitPricing';

export type ProductListProps = React.PropsWithChildren<unknown>;

export function ProductList({ children }: ProductListProps) {
   return (
      <VStack
         data-testid="list-view-products"
         spacing={2}
         align="stretch"
         width="full"
         boxSizing="border-box"
         divider={<Divider mt="0" style={{ marginTop: 0, marginBottom: 0 }} />}
      >
         {children}
      </VStack>
   );
}

export interface ProductListItemProps {
   product: ProductSearchHit;
}

export function ProductListItem({ product }: ProductListItemProps) {
   const appContext = useAppContext();

   const { price, compareAtPrice, isDiscounted, percentage } =
      useProductSearchHitPricing(product);

   const productHeadingId = `product-heading-${product.handle}`;

   const quantityAvailable = Math.max(0, product.quantity_available);

   const showProBadge = product.is_pro > 0;
   const showDiscountBadge = quantityAvailable > 0 && isDiscounted;
   const showLifetimeWarrantyBadge =
      quantityAvailable > 0 && product.lifetime_warranty;
   const showOemPartnershipBadge =
      quantityAvailable > 0 && product.oem_partnership;
   const showBadges =
      showProBadge ||
      showDiscountBadge ||
      showLifetimeWarrantyBadge ||
      showOemPartnershipBadge;

   return (
      <LinkBox as="article" aria-labelledby={productHeadingId} role="group">
         <Flex
            align="flex-start"
            py={{
               base: 4,
               xl: 7,
            }}
            px="4"
         >
            <Box
               flexGrow={0}
               flexShrink={0}
               w={{
                  base: '100px',
                  sm: '160px',
                  md: '140px',
                  lg: '160px',
               }}
            >
               {product.image_url ? (
                  <IfixitImage
                     src={product.image_url}
                     alt={product.title}
                     objectFit="contain"
                     width="180px"
                     height="180px"
                  />
               ) : (
                  <IfixitImage
                     src={placeholderImageUrl}
                     alt={product.title}
                     sizes="180px"
                  />
               )}
            </Box>
            <Stack
               direction={{
                  base: 'column',
                  xl: 'row',
               }}
               spacing={{
                  base: 1,
                  xl: 2,
               }}
               flexGrow={1}
               ml="4"
            >
               <Box flexShrink={1} w="full">
                  <Heading
                     id={productHeadingId}
                     as="h3"
                     fontSize={{
                        base: 'md',
                        lg: 'lg',
                     }}
                     mb={{
                        base: 2,
                        xl: 4,
                     }}
                     _groupHover={{ color: 'brand.500' }}
                  >
                     {product.title}
                  </Heading>
                  <Text
                     noOfLines={3}
                     fontSize={{
                        base: 'sm',
                        sm: 'sm',
                        lg: 'md',
                     }}
                     mb={{
                        base: 2,
                        xl: 4,
                     }}
                  >
                     {product.short_description}
                  </Text>
                  {(product.rating >= 4 || product.rating_count > 10) && (
                     <HStack align="center" data-testid="reviewStars">
                        <Rating value={product.rating} />
                        <Text
                           fontSize={{
                              base: 'sm',
                              lg: 'md',
                           }}
                        >
                           {product.rating_count}
                        </Text>
                     </HStack>
                  )}
                  {showBadges && (
                     <Flex
                        wrap="wrap"
                        mt="2"
                        sx={{
                           '& > *': {
                              mr: 1,
                              mt: 1,
                           },
                        }}
                     >
                        {showProBadge &&
                           ProductListItemBadge({
                              content: 'iFixit Pro',
                              colorScheme: 'orange',
                           })}
                        {showDiscountBadge &&
                           ProductListItemBadge({
                              content: `${percentage}% Off`,
                              colorScheme: 'red',
                           })}
                        {showOemPartnershipBadge &&
                           ProductListItemBadge({
                              content: product.oem_partnership!,
                              colorScheme: 'green',
                           })}
                        {showLifetimeWarrantyBadge &&
                           ProductListItemBadge({
                              content: 'Lifetime Warranty',
                              colorScheme: 'blue',
                           })}
                     </Flex>
                  )}
               </Box>
               <VStack
                  flexShrink={0}
                  align="flex-end"
                  alignSelf={{
                     base: 'flex-end',
                     xl: 'flex-start',
                  }}
               >
                  {price != null && (
                     <VStack
                        align="flex-end"
                        spacing="0"
                        mt={{
                           base: 2,
                           sm: 0,
                           md: 2,
                           lg: 0,
                        }}
                     >
                        <Text
                           color={isDiscounted ? 'red.700' : 'inherit'}
                           fontWeight="bold"
                           fontSize="xl"
                           lineHeight="1em"
                           data-testid="product-price"
                        >
                           ${price}
                        </Text>
                        {isDiscounted && (
                           <Text
                              lineHeight="1em"
                              textDecoration="line-through"
                              color="gray.400"
                              data-testid="product-compared-at-price"
                           >
                              ${compareAtPrice}
                           </Text>
                        )}
                     </VStack>
                  )}
                  <Stack
                     direction={{
                        base: 'row-reverse',
                        sm: 'column',
                        md: 'row-reverse',
                        lg: 'column',
                     }}
                     align={{
                        base: 'center',
                        sm: 'flex-end',
                        md: 'center',
                        lg: 'flex-end',
                     }}
                     spacing={{
                        base: 3,
                        sm: 2,
                     }}
                  >
                     <LinkOverlay
                        href={`${appContext.ifixitOrigin}${product.url}`}
                     >
                        <Button
                           as="div"
                           minW={{
                              base: '100px',
                              sm: '60px',
                              md: '100px',
                              lg: '60px',
                           }}
                           colorScheme="brand"
                        >
                           View
                        </Button>
                     </LinkOverlay>
                     {quantityAvailable < 10 && quantityAvailable > 0 && (
                        <Text color="gray.500" fontSize="14px">
                           Only {quantityAvailable} left in stock
                        </Text>
                     )}
                  </Stack>
               </VStack>
            </Stack>
         </Flex>
      </LinkBox>
   );
}

interface ProductListItemBadgeProps {
   content: string;
   colorScheme: ThemeTypings['colorSchemes'];
}

function ProductListItemBadge({
   content,
   colorScheme,
}: ProductListItemBadgeProps) {
   return (
      <Badge
         colorScheme={colorScheme}
         textTransform="none"
         borderRadius="lg"
         px="2.5"
         py="1"
      >
         {content}
      </Badge>
   );
}
