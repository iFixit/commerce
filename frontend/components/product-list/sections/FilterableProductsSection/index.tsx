import {
   ProductListEmptyStateIllustration,
   SearchEmptyStateIllustration,
} from '@assets/svg';
import {
   Box,
   BoxProps,
   Button,
   Flex,
   forwardRef,
   Heading,
   HStack,
   Icon,
   Link,
   Text,
   VStack,
} from '@chakra-ui/react';
import { Card } from '@components/ui';
import { IFIXIT_ORIGIN } from '@config/env';
import { getProductListTitle } from '@helpers/product-list-helpers';
import { cypressWindowLog } from '@helpers/test-helpers';
import { useLocalPreference } from '@ifixit/ui';
import {
   ProductList as TProductList,
   ProductSearchHit,
} from '@models/product-list';
import * as React from 'react';
import {
   useClearRefinements,
   useCurrentRefinements,
   useHits,
   useSearchBox,
} from 'react-instantsearch-hooks-web';
import { CurrentRefinements } from './CurrentRefinements';
import { FacetsAccordion } from './FacetsAccordion';
import { Pagination } from './Pagination';
import { ProductGrid, ProductGridItem } from './ProductGrid';
import { ProductList, ProductListItem } from './ProductList';
import { ProductViewType, Toolbar } from './Toolbar';
import { useDevicePartsItemType } from './useDevicePartsItemType';

const PRODUCT_VIEW_TYPE_STORAGE_KEY = 'productViewType';

type SectionProps = {
   productList: TProductList;
};

export function FilterableProductsSection({ productList }: SectionProps) {
   const { hits } = useHits<ProductSearchHit>();
   const [viewType, setViewType] = useLocalPreference(
      PRODUCT_VIEW_TYPE_STORAGE_KEY,
      ProductViewType.List
   );

   cypressWindowLog({ filteredProducts: hits });

   const productsContainerScrollRef = useScrollIntoViewEffect([hits]);

   const isEmpty = hits.length === 0;

   return (
      <Flex
         ref={productsContainerScrollRef}
         as="section"
         direction="column"
         align="stretch"
         mb="4"
         data-testid="filterable-products-section"
         aria-labelledby="filterable-products-section-heading"
      >
         <Heading as="h2" id="filterable-products-section-heading" srOnly>
            Products
         </Heading>
         <Toolbar
            viewType={viewType}
            productList={productList}
            onViewTypeChange={setViewType}
         />
         <CurrentRefinements />
         <HStack mt="4" align="flex-start" spacing={{ base: 0, md: 4 }}>
            <FacetCard>
               <FacetsAccordion productList={productList} />
            </FacetCard>
            <Card flex={1}>
               <ProductListEmptyState
                  productList={productList}
                  hidden={!isEmpty}
               />
               {!isEmpty && viewType === ProductViewType.Grid && (
                  <ProductGrid>
                     {hits.map((hit) => {
                        return (
                           <ProductGridItem key={hit.handle} product={hit} />
                        );
                     })}
                  </ProductGrid>
               )}
               {!isEmpty && viewType === ProductViewType.List && (
                  <ProductList>
                     {hits.map((hit) => {
                        return (
                           <ProductListItem key={hit.objectID} product={hit} />
                        );
                     })}
                  </ProductList>
               )}
               <Pagination />
            </Card>
         </HStack>
      </Flex>
   );
}

type FacetCardProps = React.PropsWithChildren<{}>;

function FacetCard({ children }: FacetCardProps) {
   return (
      <Card
         py="1.5"
         px="3"
         w="250px"
         boxShadow="md"
         maxH="calc(100vh - var(--chakra-space-4) * 2)"
         overflow="auto"
         display={{
            base: 'none',
            md: 'block',
         }}
         position="sticky"
         top="4"
      >
         {children}
      </Card>
   );
}

function useScrollIntoViewEffect(deps: React.DependencyList) {
   const ref = React.useRef<HTMLDivElement>(null);
   React.useEffect(() => {
      if (ref.current && ref.current.getBoundingClientRect().top < 0) {
         ref.current.scrollIntoView({
            behavior: 'smooth',
         });
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
   }, deps);
   return ref;
}

type EmptyStateProps = BoxProps & {
   productList: TProductList;
};

const ProductListEmptyState = forwardRef<EmptyStateProps, 'div'>(
   ({ productList, ...otherProps }, ref) => {
      const clearRefinements = useClearRefinements();

      const currentRefinements = useCurrentRefinements();
      const hasRefinements = currentRefinements.items.length > 0;

      const searchBox = useSearchBox();
      const hasSearchQuery = searchBox.query.length > 0;

      const isFiltered = hasRefinements || hasSearchQuery;

      const itemType = useDevicePartsItemType(productList);
      const title = getProductListTitle(productList, itemType);
      const encodedQuery = encodeURIComponent(searchBox.query);

      if (isFiltered) {
         return (
            <VStack
               ref={ref}
               pt="16"
               pb="20"
               px="2"
               textAlign="center"
               {...otherProps}
            >
               <Icon
                  as={SearchEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text fontSize="lg" fontWeight="bold" w="full">
                  No matching products found in {title}
               </Text>
               <Text maxW="500px" color="gray.500">
                  Try adjusting your search or filter to find what you&apos;re
                  looking for.
               </Text>
               <Link
                  maxW="500px"
                  color="brand.500"
                  href={`${IFIXIT_ORIGIN}/Search?query=${encodedQuery}`}
               >
                  Search all of iFixit for&nbsp;
                  <Text as="span" fontWeight="bold">
                     {searchBox.query}
                  </Text>
               </Link>
               <Box pt="8">
                  <Button
                     colorScheme="brand"
                     onClick={() => {
                        searchBox.clear();
                        clearRefinements.refine();
                     }}
                  >
                     Reset filters
                  </Button>
               </Box>
            </VStack>
         );
      }
      return (
         <Card
            ref={ref}
            pt="16"
            pb="20"
            borderRadius={{ base: 'none', sm: 'lg' }}
            {...otherProps}
         >
            <VStack>
               <Icon
                  as={ProductListEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text
                  fontSize="lg"
                  fontWeight="bold"
                  w="full"
                  textAlign="center"
               >
                  Empty collection
               </Text>
               <Text maxW="500px" color="gray.500" textAlign="center" px="2">
                  This collection does not have products. Try to navigate to
                  subcategories to find what you are looking for.
               </Text>
            </VStack>
         </Card>
      );
   }
);
