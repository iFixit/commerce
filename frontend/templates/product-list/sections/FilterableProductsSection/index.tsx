import {
   ProductListEmptyStateIllustration,
   SearchEmptyStateIllustration,
} from '@assets/svg/files';
import {
   Box,
   BoxProps,
   Button,
   Collapse,
   Divider,
   Flex,
   forwardRef,
   Heading,
   Icon,
   Link,
   Text,
   VStack,
} from '@chakra-ui/react';
import { ProductGrid } from '@components/common/ProductGrid';
import { ProductGridItem } from '@components/common/ProductGridItem';
import { Card } from '@components/ui';
import { filterFalsyItems } from '@helpers/application-helpers';
import { productListPath } from '@helpers/path-helpers';
import { getProductListTitle } from '@helpers/product-list-helpers';
import { useAppContext } from '@ifixit/app';
import { useLocalPreference, Wrapper } from '@ifixit/ui';
import { productPreviewFromAlgoliaHit } from '@models/components/product-preview';
import {
   ProductList as TProductList,
   ProductSearchHit,
} from '@models/product-list';
import {
   SearchQueryProvider,
   useSearchQuery,
} from '@templates/product-list/hooks/useSearchQuery';
import * as React from 'react';
import {
   useClearRefinements,
   useCurrentRefinements,
   useHits,
   useSearchBox,
} from 'react-instantsearch';
import { useDevicePartsItemType } from '../../hooks/useDevicePartsItemType';
import { CurrentRefinements } from './CurrentRefinements';
import { FacetsAccordion } from './facets/accordion';
import { Pagination } from './Pagination';
import { ProductList, ProductListItem } from './ProductList';
import { ProductViewType, Toolbar } from './Toolbar';
import { useHasAnyVisibleFacet } from './useHasAnyVisibleFacet';

const PRODUCT_VIEW_TYPE_STORAGE_KEY = 'productViewType';

type SectionProps = {
   productList: TProductList;
   algoliaSSR?: boolean;
};

export function FilterableProductsSection({
   productList,
   algoliaSSR,
}: SectionProps) {
   const { hits } = useHits<ProductSearchHit>();
   const hasAnyVisibleFacet = useHasAnyVisibleFacet(productList);
   const products = React.useMemo(
      () => filterFalsyItems(hits.map(productPreviewFromAlgoliaHit)),
      [hits]
   );
   const currentRefinements = useCurrentRefinements();
   const hasCurrentRefinements = currentRefinements.items.length > 0;
   const [viewType, setViewType] = useLocalPreference(
      PRODUCT_VIEW_TYPE_STORAGE_KEY,
      ProductViewType.List,
      (data) =>
         data === ProductViewType.List || data === ProductViewType.Grid
            ? data
            : null
   );

   const productsContainerScrollRef = useScrollIntoViewEffect([hits]);

   const isEmpty = hits.length === 0;

   if (algoliaSSR) {
      return (
         <SearchQueryProvider>
            <CurrentRefinements />
            <FacetsAccordion productList={productList} />
         </SearchQueryProvider>
      );
   }
   return (
      <Flex
         ref={productsContainerScrollRef}
         id="products"
         as="section"
         direction="column"
         align="stretch"
         data-testid="filterable-products-section"
         aria-labelledby="filterable-products-section-heading"
         my={{ base: 4, md: 6 }}
      >
         <Heading as="h2" id="filterable-products-section-heading" srOnly>
            Products
         </Heading>

         <SearchQueryProvider>
            <Wrapper>
               <Flex align="flex-start">
                  <Box
                     bg="white"
                     borderWidth="1px"
                     borderStyle="solid"
                     borderColor="gray.300"
                     borderRadius="base"
                     px="3"
                     py="1.5"
                     mr="4"
                     w="64"
                     maxH="calc(100vh - var(--chakra-space-4) * 2)"
                     overflow="auto"
                     display={{
                        base: 'none',
                        md:
                           hasAnyVisibleFacet || hasCurrentRefinements
                              ? 'block'
                              : 'none',
                     }}
                     position="sticky"
                     top="4"
                     flexGrow="0"
                  >
                     <Collapse
                        in={hasCurrentRefinements}
                        animateOpacity
                        data-testid="current-refinements"
                     >
                        <Flex mt="1.5" mb="3" wrap="wrap" align="flex-start">
                           <CurrentRefinements />
                        </Flex>
                        <Divider borderColor="gray.300" opacity="1" />
                     </Collapse>
                     <FacetsAccordion productList={productList} />
                  </Box>
                  <Flex direction="column" flex="1">
                     <Toolbar
                        viewType={viewType}
                        productList={productList}
                        onViewTypeChange={setViewType}
                     />
                     <Flex
                        bg="white"
                        direction="column"
                        mt="2"
                        overflow="hidden"
                        borderWidth="1px"
                        borderStyle="solid"
                        borderColor="gray.300"
                        borderRadius="base"
                     >
                        <ProductListEmptyState
                           productList={productList}
                           hidden={!isEmpty}
                        />
                        {!isEmpty && viewType === ProductViewType.Grid && (
                           <ProductGrid
                              data-testid="grid-view-products"
                              columns={{
                                 base: 2,
                                 sm: 3,
                                 md: 2,
                                 lg: 3,
                                 xl: 4,
                              }}
                           >
                              {products.map((product) => {
                                 return (
                                    <ProductGridItem
                                       key={product.handle}
                                       product={product}
                                    />
                                 );
                              })}
                           </ProductGrid>
                        )}
                        {!isEmpty && viewType === ProductViewType.List && (
                           <ProductList>
                              {hits.map((hit) => {
                                 return (
                                    <ProductListItem
                                       key={hit.objectID}
                                       product={hit}
                                    />
                                 );
                              })}
                           </ProductList>
                        )}
                        <Pagination />
                     </Flex>
                  </Flex>
               </Flex>
            </Wrapper>
         </SearchQueryProvider>
      </Flex>
   );
}

function useScrollIntoViewEffect(deps: React.DependencyList) {
   const ref = React.useRef<HTMLDivElement>(null);
   React.useEffect(() => {
      if (ref.current && ref.current.getBoundingClientRect().top < 0) {
         ref.current.scrollIntoView({
            behavior: 'smooth',
         });
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
   }, deps);
   return ref;
}

type EmptyStateProps = BoxProps & {
   productList: TProductList;
};

const ProductListEmptyState = forwardRef<EmptyStateProps, 'div'>(
   ({ productList, ...otherProps }, ref) => {
      const { setSearchQuery } = useSearchQuery();
      const clearRefinements = useClearRefinements({ excludedAttributes: [] });

      const currentRefinements = useCurrentRefinements();
      const hasRefinements = currentRefinements.items.length > 0;

      const searchBox = useSearchBox();
      const hasSearchQuery = searchBox.query.length > 0;

      const isFiltered = hasRefinements || hasSearchQuery;

      const itemType = useDevicePartsItemType(productList);
      const title = getProductListTitle(
         {
            title: productList.title,
            type: productList.type,
         },
         itemType
      );
      const encodedQuery = encodeURIComponent(searchBox.query);

      const ancestors = productList.ancestors;
      const parentCategory = ancestors[ancestors.length - 1];

      const appContext = useAppContext();

      if (isFiltered) {
         return (
            <VStack
               ref={ref}
               pt="16"
               pb="20"
               px="2"
               textAlign="center"
               {...otherProps}
               data-testid="product-list-no-results"
            >
               <Icon
                  as={SearchEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text fontSize="lg" fontWeight="bold" w="full">
                  No matching products found in {title}
               </Text>
               <Text maxW="500px" color="gray.500">
                  Try adjusting your search or filter to find what you&apos;re
                  looking for.
               </Text>
               <Link
                  maxW="500px"
                  color="brand.500"
                  href={`${appContext.ifixitOrigin}/Search?query=${encodedQuery}`}
               >
                  Search all of iFixit for&nbsp;
                  <Text as="span" fontWeight="bold">
                     {searchBox.query}
                  </Text>
               </Link>
               <Box pt="8">
                  <Button
                     colorScheme="brand"
                     onClick={() => {
                        setSearchQuery('');
                        clearRefinements.refine();
                     }}
                  >
                     Reset filters
                  </Button>
               </Box>
            </VStack>
         );
      }
      return (
         <Card
            ref={ref}
            pt="16"
            pb="20"
            borderRadius={{ base: 'none', sm: 'lg' }}
            sx={{
               a: {
                  color: 'brand.500',
                  transition: 'color 300ms',
                  '&:hover': {
                     color: 'brand.600',
                  },
               },
            }}
            {...otherProps}
         >
            <VStack>
               <Icon
                  as={ProductListEmptyStateIllustration}
                  boxSize="200px"
                  opacity="0.8"
               />
               <Text
                  fontSize="lg"
                  fontWeight="bold"
                  w="full"
                  textAlign="center"
               >
                  Empty collection
               </Text>
               <Text maxW="500px" color="gray.500" textAlign="center" px="2">
                  This collection does not have products.
               </Text>
               {parentCategory && (
                  <Link href={productListPath(parentCategory)}>
                     Return to {parentCategory.title}
                  </Link>
               )}
            </VStack>
         </Card>
      );
   }
);
