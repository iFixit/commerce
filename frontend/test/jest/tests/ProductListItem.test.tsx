import { render, screen } from '@testing-library/react';
import { ProductListItem } from '@components/product-list/sections/FilterableProductsSection/ProductList';
import * as ProductSearch from '@components/product-list/sections/FilterableProductsSection/useProductSearchHitPricing';
import { mockProduct } from 'test/jest/__mocks__/mockProduct';

jest.mock('@ifixit/app');

describe('ProductListItem', () => {
   beforeEach(() => {
      // @ts-ignore
      ProductSearch.useProductSearchHitPricing = jest.fn().mockReturnValue({
         price: 5,
         compareAtPrice: 5,
         isDiscounted: 5,
         percentage: 5,
      });
   });

   afterEach(() => {
      jest.clearAllMocks();
   });

   it('renders and matches the snapshot', () => {
      // @ts-ignore
      const { asFragment } = render(<ProductListItem product={mockProduct} />);
      const shortDescription = screen.getByText(
         'Replace a dead or malfunctioning model EB-BG96aasd5ABE battery in a Samsung Galaxy S9 Plus smartphone.'
      );

      (expect(shortDescription) as any).toBeInTheDocument();
      (expect(asFragment()) as any).toMatchSnapshot();
   });

   it('renders without the review stars', () => {
      // We don't render the stars if the rating <= 4 AND rating_count < 10
      mockProduct.rating = 3.5;
      mockProduct.rating_count = 9;

      // @ts-ignore
      render(<ProductListItem product={mockProduct} />);

      const reviewStars = screen.queryByTestId('reviewStars');
      (expect(reviewStars) as any).not.toBeInTheDocument();
   });

   it('renders with the review stars', () => {
      // We render the stars if the rating >= 4 OR rating_count > 10

      // If rating < 4; count > 10
      mockProduct.rating = 1;
      mockProduct.rating_count = 15;
      // @ts-ignore
      const { rerender } = render(<ProductListItem product={mockProduct} />);
      const reviewStars = screen.queryByTestId('reviewStars');
      (expect(reviewStars) as any).toBeInTheDocument();

      // If rating > 4; count < 10
      mockProduct.rating = 5;
      mockProduct.rating_count = 5;
      // @ts-ignore
      rerender(<ProductListItem product={mockProduct} />);
      (expect(reviewStars) as any).toBeInTheDocument();

      // If rating > 4; count > 10
      mockProduct.rating = 5;
      mockProduct.rating_count = 15;
      // @ts-ignore
      rerender(<ProductListItem product={mockProduct} />);
      (expect(reviewStars) as any).toBeInTheDocument();
   });
});
