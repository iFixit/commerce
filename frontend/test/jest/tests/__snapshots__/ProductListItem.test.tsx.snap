// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListItem renders and matches the snapshot 1`] = `
<DocumentFragment>
  <article
    aria-labelledby="product-heading-galaxy-s9-plus-replacement-battery"
    class="chakra-linkbox css-hng3kb"
    role="group"
  >
    <div
      class="css-19p2217"
    >
      <div
        class="css-1vobfjw"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTgwIiBoZWlnaHQ9IjE4MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Galaxy S9+ Battery"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
          />
          <noscript />
        </div>
      </div>
      <div
        class="chakra-stack css-azp3jj"
      >
        <div
          class="css-1kx59gv"
        >
          <h3
            class="chakra-heading css-duzb3d"
            id="product-heading-galaxy-s9-plus-replacement-battery"
          >
            Galaxy S9+ Battery
          </h3>
          <p
            class="chakra-text css-1emk585"
          >
            Replace a dead or malfunctioning model EB-BG96aasd5ABE battery in a Samsung Galaxy S9 Plus smartphone.
          </p>
          <div
            class="chakra-stack css-84zodg"
            data-testid="reviewStars"
          >
            <div
              class="chakra-stack css-193rmy8"
            >
              <svg
                class="css-6p0vz0"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                />
              </svg>
              <svg
                class="css-6p0vz0"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                />
              </svg>
              <svg
                class="css-6p0vz0"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                />
              </svg>
              <svg
                class="css-6p0vz0"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                />
              </svg>
              <div
                class="css-k008qs"
              >
                <svg
                  class="css-15nhrxw"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 576 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 0c-11.4 0-22.8 5.9-28.7 17.8L194 150.2 47.9 171.4c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.1 23 46 46.4 33.7L288 439.6V0z"
                  />
                </svg>
                <svg
                  class="css-1g2soo"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 576 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 0c-11.4 0-22.8 5.9-28.7 17.8L194 150.2 47.9 171.4c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.1 23 46 46.4 33.7L288 439.6V0z"
                  />
                </svg>
              </div>
            </div>
            <p
              class="chakra-text css-xh8f59"
            >
              18
            </p>
          </div>
          <div
            class="css-14us8ue"
          >
            <span
              class="chakra-badge css-1l0yll"
            >
              5% Off
            </span>
          </div>
        </div>
        <div
          class="chakra-stack css-1x25lam"
        >
          <div
            class="chakra-stack css-1wwkyaq"
          >
            <p
              class="chakra-text css-12jr671"
              data-testid="product-price"
            >
              $5
            </p>
            <p
              class="chakra-text css-m9wb1i"
              data-testid="product-compared-at-price"
            >
              $5
            </p>
          </div>
          <div
            class="chakra-stack css-1dz9rto"
          >
            <a
              class="chakra-linkbox__overlay css-kvj8dg"
              href="https://www.cominor.com/Store/Android/Galaxy-S9-Battery/IF404-007?o=4"
            >
              <div
                class="chakra-button css-1xl7ue1"
              >
                View
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
